nvflare/widgets/fed_event.py

```diff
@@ -109,29 +116,61 @@ def _receive(self, topic: str, request: Shareable, fl_ctx: FLContext) -> Shareab
         return make_reply(ReturnCode.OK)
 
     def _post(self):
-        sleep_time = 0.1
+        """
+        During ABOUT_TO_END_RUN, sleep_time is 0 and system will flush
+         in_events by firing events without delay.
+        During END_RUN, system will wait for self.grace_period, even the queue is empty,
+        so any new item can be processed.
+        However, since the system does not guarantee the receiving side of _post is still
+        alive, we catch the exception and show warning messages to users if events can not
+        be handled by receiving side.
+        """
+        sleep_time = self.regular_interval
+        countdown = self.grace_period
         while True:
             time.sleep(sleep_time)
-            if self.asked_to_stop or self.abort_signal.triggered:
+            if self.abort_signal.triggered:
                 break
-
-            with self.in_lock:
-                if len(self.in_events) <= 0:
+            n = len(self.in_events)
+            if n > 0:
+                if self.asked_to_flush:
+                    sleep_time = 0
+                else:
+                    sleep_time = self.regular_interval
+                with self.in_lock:
+                    event_to_post = self.in_events.pop(0)
+            elif self.asked_to_stop:
+                # the queue is empty and we are asked to stop.  Give
+                # it self.grace_period seconds to wait, then
+                # exit.
+                if countdown < 0:
+                    break
+                else:
+                    countdown = countdown - 1
+                    time.sleep(1)
                     continue
-                event_to_post = self.in_events.pop(0)
-                assert isinstance(event_to_post, Shareable)
-
-            if self.asked_to_stop or self.abort_signal.triggered:
-                break
+            else:
+                sleep_time = min(sleep_time * 2, 1)
+                continue
 
+            assert isinstance(event_to_post, Shareable)
```

Comment:
I think we don't need this statement, as self.in_events is added by this class.

Reply:
This is from original codes.  But I agree this assert is not needed.